File: bindings/matlab/+proven/SafeString.m
% SPDX-License-Identifier: PMPL-1.0-or-later
% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%
% SafeString - Safe string operations via libproven FFI.
%
% All computation is performed in the Idris 2 core via the Zig FFI bridge.
% This class is a thin wrapper; it does NOT reimplement any logic.

classdef SafeString
    % SafeString  Formally verified text operations that handle encoding safely.
    %
    % Methods return [] on error, or a char/logical value on success.
    %
    % Example:
    %   proven.LibProven.init();
    %   html = proven.SafeString.escapeHtml('<script>alert(1)</script>');
    %   % html == '&lt;script&gt;alert(1)&lt;/script&gt;'

    methods (Static)

        function result = isValidUtf8(str)
            % ISVALIDUTF8  Check if bytes are valid UTF-8.
            %
            %   tf = proven.SafeString.isValidUtf8(str)
            %
            % Parameters:
            %   str - character vector or uint8 array.
            %
            % Returns logical true/false or [] on error.
            proven.LibProven.ensureLoaded();
            data = uint8(str);
            r = calllib(proven.LibProven.libName(), ...
                        'proven_string_is_valid_utf8', data, numel(data));
            if r.status == 0
                result = logical(r.value);
            else
                result = [];
            end
        end

        function result = escapeSql(str)
            % ESCAPESQL  Escape string for SQL (single quotes).
            %
            %   escaped = proven.SafeString.escapeSql(str)
            %
            % Returns escaped char vector or [] on error.
            % Note: Prefer parameterized queries over string escaping.
            proven.LibProven.ensureLoaded();
            data = uint8(str);
            r = calllib(proven.LibProven.libName(), ...
                        'proven_string_escape_sql', data, numel(data));
            if r.status == 0 && r.length > 0
                result = char(r.value(1:r.length));
                proven.LibProven.freeString(r.value);
            elseif r.status == 0
                result = '';
            else
                result = [];
            end
        end

        function result = escapeHtml(str)
            % ESCAPEHTML  Escape string for HTML (prevents XSS).
            %
            %   escaped = proven.SafeString.escapeHtml(str)
            %
            % Returns escaped char vector or [] on error.
            proven.LibProven.ensureLoaded();
            data = uint8(str);
            r = calllib(proven.LibProven.libName(), ...
                        'proven_string_escape_html', data, numel(data));
            if r.status == 0 && r.length > 0
                result = char(r.value(1:r.length));
                proven.LibProven.freeString(r.value);
            elseif r.status == 0
                result = '';
            else
                result = [];
            end
        end

        function result = escapeJs(str)
            % ESCAPEJS  Escape string for JavaScript string literals.
            %
            %   escaped = proven.SafeString.escapeJs(str)
            %
            % Returns escaped char vector or [] on error.
            proven.LibProven.ensureLoaded();
            data = uint8(str);
            r = calllib(proven.LibProven.libName(), ...
                        'proven_string_escape_js', data, numel(data));
            if r.status == 0 && r.length > 0
                result = char(r.value(1:r.length));
                proven.LibProven.freeString(r.value);
            elseif r.status == 0
                result = '';
            else
                result = [];
            end
        end

    end
end

File: bindings/swift/Sources/Proven/SafeString.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe string operations delegated to libproven FFI.
///
/// All escaping and validation is performed by the formally verified
/// Idris 2 core via the Zig FFI bridge.

import CProven

public enum SafeString {
    /// Check if bytes are valid UTF-8.
    public static func isValidUtf8(_ data: [UInt8]) -> Result<Bool, ProvenError> {
        let result = data.withUnsafeBufferPointer { buffer in
            proven_string_is_valid_utf8(buffer.baseAddress, buffer.count)
        }
        if let error = ProvenError.fromStatus(result.status) {
            return .failure(error)
        }
        return .success(result.value)
    }

    /// Escape a string for safe SQL interpolation.
    /// Note: Prefer parameterized queries over string escaping.
    public static func escapeSql(_ value: String) -> Result<String, ProvenError> {
        withStringBytes(value) { ptr, len in
            consumeStringResult(proven_string_escape_sql(ptr, len))
        }
    }

    /// Escape a string for safe HTML insertion.
    public static func escapeHtml(_ value: String) -> Result<String, ProvenError> {
        withStringBytes(value) { ptr, len in
            consumeStringResult(proven_string_escape_html(ptr, len))
        }
    }

    /// Escape a string for safe JavaScript string literal insertion.
    public static func escapeJs(_ value: String) -> Result<String, ProvenError> {
        withStringBytes(value) { ptr, len in
            consumeStringResult(proven_string_escape_js(ptr, len))
        }
    }
}
